File: Data Structures & Algorithms/longest-substring-without-duplicates/submission-0.js
class Solution {
    /**
     * @param {string} s
     * @return {number}
     */
    lengthOfLongestSubstring(s) {
        let hash = new Map()
        let left = 0
        let len = 0

        for(let right = left; right < s.length; right++){
            hash.set(s[right], (hash.get(s[right]) || 0) + 1)

            while(hash.get(s[right])>1){
                hash.set(s[left], hash.get(s[left]) - 1)

                if(hash.get(s[left]) === 0){
                    hash.delete(s[left])
                }
                left ++
            }
            len = Math.max(len, right - left + 1)
        }
        return len
    }
}

File: Data Structures & Algorithms/longest-substring-without-duplicates/submission-0.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        maxLen = 0

        left = 0
        window = {}
        for right in range(len(s)):
            window[s[right]] = window.get(s[right], 0) + 1

            while window[s[right]] > 1 :
                window[s[left]] -=1
            
                if window[s[left]] == 0:
                    del window[s[left]]
                left +=1
            maxLen = max(maxLen, right - left + 1)
        return maxLen



        